ampd/src/sui/verifier.rs

```diff
@@ -0,0 +1,194 @@
+use move_core_types::language_storage::StructTag;
+use serde::Deserialize;
+use sui_json_rpc_types::{SuiEvent, SuiTransactionBlockResponse};
+use sui_types::base_types::{ObjectID, SuiAddress};
+
+use crate::handlers::sui_verify_msg::Message;
+use crate::types::Hash;
+
+const CONTRACT_CALL_EVENT: &str = "gateway::ContractCall";
+
+// TODO: update after Sui gateway event finalization
+#[derive(Deserialize)]
+struct ContractCall {
+    pub source_id: SuiAddress,
+    pub destination_chain: String,
+    pub destination_address: String,
+    pub payload_hash: Hash,
+}
+
+// event type is in the form of: <gateway_address>::gateway::ContractCall
+fn call_contract_type(gateway_address: &ObjectID) -> StructTag {
+    format!("{}::{}", gateway_address, CONTRACT_CALL_EVENT)
+        .as_str()
+        .parse()
+        .expect("failed to parse struct tag")
+}
+
+impl PartialEq<&Message> for &SuiEvent {
+    fn eq(&self, msg: &&Message) -> bool {
+        if self.type_ != call_contract_type(&self.package_id) {
+            return false;
+        }
+
+        serde_json::from_value::<ContractCall>(self.parsed_json.clone()).unwrap();
+
+        match serde_json::from_value::<ContractCall>(self.parsed_json.clone()) {
+            Ok(contract_call) => {
+                contract_call.source_id == msg.source_address
+                    && contract_call.destination_chain == msg.destination_chain.to_string()
+                    && contract_call.destination_address == msg.destination_address
+                    && contract_call.payload_hash == msg.payload_hash
+            }
+            _ => false,
+        }
+    }
+}
+
+fn get_event<'a>(
+    gateway_address: &SuiAddress,
+    transaction_block: &'a SuiTransactionBlockResponse,
+    event_seq: u64,
+) -> Option<&'a SuiEvent> {
+    if let Some(events) = transaction_block.events.as_ref() {
```

Comment:
probably cleaner to go with `transaction_block.events.as_ref().map(|events|...`, then you don't need the else case